### ISSUE TITLE
find_on_path distribution finder can cause "maximum recursion depth exceeded in comparison"

### ISSUE THREAD

**Issue by cjerdonek** [pypa/setuptools#1708]
A few days ago someone filed an issue on pip's tracker that looks like a setuptools issue:
https://github.com/pypa/pip/issues/6288

Basically, it looks like the issue is that `find_on_path()` (the "distribution finder" corresponding to `pkgutil.ImpImporter` and `FileFinder`) can get into a situation where it's calling itself in an (infinite) recursive loop. Also, I believe this is happening in the `find_on_path(only=False)` case, rather than `only=True`.

**Comment by cjerdonek** [comment]
I wonder if this could be caused by a path visited by `find_on_path()` resolving as a symlink to a path higher in the directory hierarchy already visited by `find_on_path()`. It looks like `find_on_path()` function does resolve symlinks here in the first line (the `_normalize_cached()` function does end up calling `os.path.realpath()`):
https://github.com/pypa/setuptools/blob/702f2d9127adc6fb6922a2c164b0a9a2856f203a/pkg_resources/__init__.py#L2018-L2020

**Comment by Rajskc** [comment]
Hey @pganssle  . Can you please let me know what can be done to solve this bug?

**Comment by pganssle** [comment]
@Rajskc I'm not actually sure. If either you or @cjerdonek can provide a minimal test case that reproduces the issue, it would be a good start towards solving the issue. Even better if you could make a PR with a failing test case (marked as `xfail`).

**Comment by Rajskc** [comment]
@cjerdonek  Any suggestion would be very helpful as I am not quite familiar with this stuff.

**Comment by Rajskc** [comment]
@cjerdonek  Hey! I haven't been able to work on this till now. Can you please help me here. I understand that it is getting stuck the loop below, does increasing the sys.setrecursionlimit  to 1500 help here and is it a good method to solve this.
File "c:\users\m phanidhar\anaconda3\lib\site-packages\pip\_vendor\pkg_resources\__init__.py", line 1968, in find_on_path
    for dist in factory(fullpath):
 [Previous line repeated 960 more times]

**Comment by cjerdonek** [comment]
I wouldn't recommend increasing the recursion limit..

**Comment by Rajskc** [comment]
Okay I will create a PR as suggested above.

**Comment by Rajskc** [comment]
hey @pganssle  I am trying to create a pull request as you suggested. But the code is different in my system compared to the gitversion here. I have installed the latest setup tools version using pip. What can i do to solve this?

**Comment by pganssle** [comment]
@Rajskc At the moment, I think we're still blocked on _reproducing_ the problem. The first step is preparing a *minimal* working example of a way for someone else to reproduce the problem. Here is [a blog post on crafting minimal reports](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports), and here's [Stack Overflow's guide](https://stackoverflow.com/help/mcve). Even if you already have the solution, I have found that most of the time I spend working on PRs that are eventually merged is spent writing tests, and tests are basically MWEs that run automatically.

When you are ready to make a PR, you should be working in the git repository, so you would need to fork the repo, then clone it locally, then make changes in a new branch. [Here is a document on how to make a PR](https://gist.github.com/rsp/8e565895df24c46ee80cfaac68d05e64). You don't necessarily have to install the git version on your local system, because if you have added a test, you can install and run `tox` from the local environment.

For more information about the *specifics* of contributing to `setuptools`, see our [contributing documentation](https://setuptools.readthedocs.io/en/latest/developer-guide.html) (which I realize could be better for beginners).

Thanks for your interest in helping with this. Please note that the most important thing is submitting a MWE, so that even if you get frustrated with submitting a PR, someone else can pick up where you left off.

**Comment by tronghieu1987** [comment]
Hello,

I am experiencing this issue and desperately searching for a solution. Is there any progress on this? Or could I do anything to reproduce  the error? 

I saw a MWE was asked for, but I am not sure what a "reproducing this problem" actually means. I run into this error when doing 'pip install ' with any package.

Thank you,

**Comment by jaraco** [comment]
> I am not sure what a "reproducing this problem" actually means. I run into this error when doing 'pip install ' with any package.

The problem is, when the maintainers here run `pip install`, we don't encounter the issue... so we're kinda stuck investigating. I've tried running `pip install service-identity` on my Windows machine and it completes just fine. This all suggests to me there's another issue at play here. We need to figure out what that is. Maybe you have symlinks installed. Maybe you're in a directory that has settings configured. Maybe you have environment variables set. Maybe you're behind a weird firewall. Maybe you're running on an unprivileged account. We need to eliminate all these variables and hone in on what are the factors that lead to the failure you've encountered. Now that we have two people encountering the issue, perhaps we could compare environments to ascertain what things might be in common.

cjerdonek has done some investigation trying to imagine what _might_ be going wrong, but we don't have a good way to confirm (or to ensure this doesn't happen to someone else).

For next steps, I recommend trying another machine. If you're on a work machine, find a personal machine (or vice versa) to try the same operation. Or try installing a clean version of Python (maybe try 3.7 instead of 3.6). Or find a friend and see if they encounter the same problem. Alternately, install a clean version of Windows into a VM and try that. Or even better, if you can reproduce the issue outside of Windows, that will vastly expand the audience of maintainers that will be able to help track down the issue.

Failing all of that, I recommend digging into the code when the failure occurs (add print statements or use [pdb](https://docs.python.org/3/library/pdb.html) to break into the code and inspect the state. You'll need to study the code enough to understand what it's _trying_ to do to know when the behavior diverges from that expectation.

Best of luck.

**Comment by JVillella** [comment]
I ran into this problem today. My setup is probably different than the OP, but the errors appear to be the same.

I had a docker container w/ the following,

```dockerfile
RUN cd /usr/local/lib/python3.7/dist-packages && python /src/setup.py develop
```

Purpose was so that the `.egg-link` doesn't live under `/src` and I can mount a src volume. Not doing this would overwrite the `.egg-link` when mounting.

Then, once in the container, trying to run any `pip install ...` returns the `find_path()` recursion failure. The problem was that I was using `dist-packages`, should have been `site-packages` (the latter of which didn't initially exist in the container). Here's what fixed the problem,

```dockerfile
ENV PYTHONPATH=/src
COPY . /src
RUN cd /usr/local/lib/python3.7 && \
    mkdir site-packages && \
    cd site-packages && \
    python /src/setup.py develop
WORKDIR /src
```

**Comment by phonokoye** [comment]
Hello, Is anyone still on this issue, I am volunteering to create a MRE (Minimum Reproducible Example), I sort of temporarily have the time for this. If anyone would be willing to guide me through what is needed. As a first start, I got the following logs by modifying ssl.py (Probably not recommended) on Python 3.10.5, pip 25.

```
@verify_mode.setter
    def verify_mode(self, value):
        print('Debug Statement: verify_mode set to', value)
        super(SSLContext, SSLContext).verify_mode.__set__(self, value)
```

I ran `pip install colorama --upgrade`

Requirement already satisfied: colorama in c:\python\python310\lib\site-packages (0.4.6)

```
Requirement already satisfied: colorama in c:\python\python310\lib\site-packages (0.4.6)
Debug Statement: verify_mode set to VerifyMode.CERT_REQUIRED
Debug Statement: verify_mode set to VerifyMode.CERT_REQUIRED
Debug Statement: verify_mode set to VerifyMode.CERT_REQUIRED
.
.
.
Debug Statement: verify_mode set to VerifyMode.CERT_REQUIRED
```

**Comment by phonokoye** [comment]
I have not gotten any reply, but before I let go, I wanted to add a piece of information that might be useful to anyone who finally takes on this issue.

1. I notice that whenever such errors occur, I can create a new virtual environment and using pip in there works without errors.
2. I have a lot of packages installed in my global environment and I do not know if this might have anything to do with it. Maybe pip has some trouble working in an environment when it has more than a specific  number of packages. I wouldn't know how to test this and that is why it would be great to have some expert on this give me some feedback

With that being said, I hope that this info might be useful in pinpointing the source of the errors.

**Comment by jaraco** [comment]
pkg_resources is deprecated and being removed, making this issue obsolete.
